cmd/snap-confine/mount-support.c

```diff
@@ -509,6 +511,38 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 	// mount table and software inspecting the mount table may become confused.
 	sc_must_snprintf(src, sizeof src, "%s/proc", SC_HOSTFS_DIR);
 	sc_do_umount(src, UMOUNT_NOFOLLOW | MNT_DETACH);
+	// Detach the both views of /writable: the one from hostfs and the one
+	// directly visible in /writable. Interfaces don't grant access to this
+	// directory and it has a large duplicated view of many mount points.
+	// Note that this is only applicable to ubuntu-core systems.
+	sc_detach_views_of_writable(config->distro, config->normal_mode);
+}
+
+static void sc_detach_views_of_writable(sc_distro distro, bool normal_mode)
+{
+	const char *writable_dir = "/writable";
+	const char *hostfs_writable_dir = "/var/lib/snapd/hostfs/writable";
+
+	// Writable only exists on ubuntu-core.
+	if (distro == SC_DISTRO_CLASSIC) {
+		return;
+	}
+	// On all core distributions we see /var/lib/snapd/hostfs/writable that
+	// exposes writable, with a structure specific to ubuntu-core. 
+	debug("detaching %s", hostfs_writable_dir);
+	sc_do_mount("none", hostfs_writable_dir, NULL,
+		    MS_REC | MS_PRIVATE, NULL);
+	sc_do_umount(hostfs_writable_dir, UMOUNT_NOFOLLOW | MNT_DETACH);
+
+	// On ubuntu-core 16, when the executed snap uses core as base we also see
+	// the /writable that we directly inherited from the initial mount
+	// namespace.
+	if (distro == SC_DISTRO_CORE16 && !normal_mode) {
```

Comment:
Obviously not to be considered in this PR, but it'd be great if we had more descriptive name for `normal_mode`. Each time I see it, I need to go back and check what it means.

Reply:
I was hoping to be able to remove the legacy mode and have no more mode at all but this idea seems futile. Perhaps you have some suggestions on where and how to describe those. In my eyes those are:

normal mode: rootfs constructed from base snap _only_
legacy mode: rootfs constructed from booted filesystem with anything visible there